File: jmlparser-jml-tools/src/test/java/com/github/jmlparser/smt/SmtTest.java
package com.github.jmlparser.smt;

import com.github.javaparser.JavaParser;
import com.github.javaparser.ParseResult;
import com.github.javaparser.ParserConfiguration;
import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.Node;
import com.github.javaparser.ast.expr.Expression;
import com.github.javaparser.symbolsolver.JavaSymbolSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.ClassLoaderTypeSolver;
import com.github.jmlparser.smt.model.SExpr;
import com.google.common.truth.Truth;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;
import org.yaml.snakeyaml.Yaml;

import java.io.IOException;
import java.io.InputStream;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.fail;

/**
 * @author Alexander Weigl
 * @version 1 (14.06.22)
 */
public class SmtTest {
    private final JavaParser parser;
    private final Node parent;

    {
        ParserConfiguration config = new ParserConfiguration();
        config.setSymbolResolver(new JavaSymbolSolver(new ClassLoaderTypeSolver(ClassLoader.getSystemClassLoader())));
        parser = new JavaParser(config);

        ParseResult<CompilationUnit> r = parser.parse(getClass().getResourceAsStream("Test.java"));
        if (!r.isSuccessful()) {
            r.getProblems().forEach(System.err::println);
            fail("Error during parsing");
        }
        parent = r.getResult().get().getType(0);
    }

    @TestFactory
    public Stream<DynamicTest> smtTranslation() throws IOException {
        try (InputStream inputStream = getClass().getResourceAsStream("expr.yaml")) {
            Yaml yaml = new Yaml();
            List<Map<String, Object>> obj = yaml.load(inputStream);
            return obj.stream().map(m -> {
                String a = (String) m.get("expr");
                String result = (String) m.get("result");
                String resultInt = (String) m.get("resultInt");
                String resultBv = (String) m.get("resultBv");
                return DynamicTest.dynamicTest(a, () -> {
                    if (resultInt != null)
                        smtTranslation(a, resultInt, true);
                    if (resultBv != null)
                        smtTranslation(a, resultBv, false);
                    if (result != null)
                        smtTranslation(a, result, false);
                });
            });
        }
    }

    void smtTranslation(String input, String expected, boolean useInt) {
        ParseResult<Expression> e = parser.parseJmlExpression(input);
        if (!e.isSuccessful()) {
            e.getProblems().forEach(System.err::println);
            fail("Error during parsing");
        }
        Expression expr = e.getResult().get();
        expr.setParentNode(parent);
        SmtQuery smtLog = new SmtQuery();
        SExpr actualExpr = SMTFacade.toSmt(expr, smtLog, useInt);

        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);
        smtLog.appendTo(pw);
        actualExpr.appendTo(pw);
        String actual = sw.toString();
        Truth.assertThat(actual.trim()).isEqualTo(expected.trim());
    }

}

File: tools/smt/src/test/kotlin/SmtTest.kt
import com.github.javaparser.JavaParser
import com.github.javaparser.ParserConfiguration
import com.github.javaparser.Problem
import com.github.javaparser.ast.Node
import com.github.javaparser.ast.expr.Expression
import com.github.javaparser.symbolsolver.JavaSymbolSolver
import com.github.javaparser.symbolsolver.resolution.typesolvers.ClassLoaderTypeSolver
import com.google.common.truth.Truth
import io.github.jmltoolkit.smt.SMTFacade
import io.github.jmltoolkit.smt.SmtQuery
import io.github.jmltoolkit.smt.model.SExpr
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import org.yaml.snakeyaml.Yaml
import java.io.PrintWriter
import java.io.StringWriter
import java.util.function.Consumer

data class ExprTestCase(
    val expr: String,
    val result: String? = null,
    val resultBv: String? = null,
    val resultInt: String? = null,
    val disabled: Boolean = false
) {
    constructor(m: Map<String, Any?>)
            : this(
        expr = m["expr"] as String,
        result = m["result"] as String?,
        resultInt = m["resultInt"] as String?,
        resultBv = m["resultBv"] as String?,
        disabled = m["disabled"] as? Boolean ?: false
    )

}

/**
 * @author Alexander Weigl
 * @version 1 (14.06.22)
 */
class SmtTest {
    private val parser: JavaParser
    private val parent: Node

    init {
        val config = ParserConfiguration()
        config.setSymbolResolver(JavaSymbolSolver(ClassLoaderTypeSolver(ClassLoader.getSystemClassLoader())))
        parser = JavaParser(config)

        val r = parser.parse(javaClass.getResourceAsStream("Test.java"))
        if (!r.isSuccessful) {
            r.problems.forEach(Consumer { x: Problem? -> System.err.println(x) })
            Assertions.fail<Any>("Error during parsing")
        }
        parent = r.result.get().getType(0)
    }


    @TestFactory
    fun smtTranslation(): List<DynamicTest> {
        javaClass.getResourceAsStream("expr.yaml").use { inputStream ->
            val yaml = Yaml()
            val obj: List<Map<String, Any>> = yaml.load(inputStream)
            return obj
                .map { ExprTestCase(it) }
                .filter { !it.disabled }
                .flatMap {
                    listOf(
                        DynamicTest.dynamicTest(it.expr + " int") {
                            it.resultInt?.let { r -> smtTranslation(it.expr, r, true) }
                        },
                        DynamicTest.dynamicTest(it.expr + " bv") {
                            it.resultBv?.let { r -> smtTranslation(it.expr, r, false) }
                        },
                        DynamicTest.dynamicTest(it.expr + " _") {
                            it.result?.let { r -> smtTranslation(it.expr, r, true) }
                        }
                    )
                }
        }
    }

    fun smtTranslation(input: String?, expected: String, useInt: Boolean) {
        val e = parser.parseJmlExpression<Expression>(input)
        if (!e.isSuccessful) {
            e.problems.forEach(Consumer { x: Problem? -> System.err.println(x) })
            Assertions.fail<Any>("Error during parsing")
        }
        val expr = e.result.get()
        expr.setParentNode(parent)
        val smtLog = SmtQuery()
        val actualExpr: SExpr = SMTFacade.toSmt(expr, smtLog, useInt)

        val sw = StringWriter()
        val pw = PrintWriter(sw)
        smtLog.appendTo(pw)
        actualExpr.appendTo(pw)
        val actual = sw.toString()
        Truth.assertThat(actual.trim { it <= ' ' }).isEqualTo(expected.trim { it <= ' ' })
    }
}
